docs: Add details on using Skia (#20171)
unoplatform/uno: doc/articles/features/using-skia-rendering.md

### Publish-time switching

When building in the `SkiaRenderer` node, an application compiles against "reference" versions of `Uno.UI`, `Uno.UI.Dispatching`, `Uno.UI.Composition`, `Uno`, and `Uno.Foundation`. When the application is being packaged, `Uno.UI` and `Uno.UI.Composition` are switched to the Skia-compatible versions. The `Uno.UI.Dispatching`, `Uno`, and `Uno.Foundation` are switched to their corresponding target platform versions.

By doing so, any use of the APIs provided by `Uno.UI.Dispatching`, `Uno`, and `Uno.Foundation` is automatically redirected to the proper platform support, for instance, redirecting `GeoLocator` to use the proper APIs provided by the underlying platform.

### Implications for iOS/Android class libraries

At this time, NuGet packages that provide UI features on iOS/Android may not be consumed by both native and Skia renderer projects under certain conditions.

The reason for this limitation is caused by the fact that Native renderers expose a different API set for `Microsoft.UI.Xaml` types, therefore causing incompatibility issues when apps use such a library.

Here are the different scenarios:

> - Given a library that uses `net9.0-ios` or `net9.0-android`, which does not have the `SkiaRender` set as an `UnoFeature`, but uses platform conditional code with `#if` blocks:
> 
>   - This package is only usable by both native and Skia apps if it also provides a `net9.0` TFM. In this case, for a native app, nothing changes.
>   - For a `SkiaRenderer` enabled app, the `net9.0` variant of the library will be used and will not offer iOS/Android specific conditional code, but any code that uses Uno Platform provided APIs will work properly.
> 
> - Given a library that uses `net9.0-ios` or `net9.0-android`, which does have the `SkiaRender` set as an `UnoFeature`, but uses platform conditional code with `#if` blocks:

Reviewer: ```suggestion
- Given a library that uses `net9.0-ios` or `net9.0-android`, which does not have the `SkiaRenderer` set as an `UnoFeature`, but uses platform conditional code with `#if` blocks:

  - This package is only usable by both native and Skia apps if it also provides a `net9.0` TFM. In this case, for a native app, nothing changes.
  - For a `SkiaRenderer` enabled app, the `net9.0` variant of the library will be used and will not offer iOS/Android specific conditional code, but any code that uses Uno Platform provided APIs will work properly.

- Given a library that uses `net9.0-ios` or `net9.0-android`, which does have the `SkiaRenderer` set as an `UnoFeature`, but uses platform conditional code with `#if` blocks:
```

> ## Architecture
> 
> In order to accomodate for the inclusion of Skia rendering for all platforms, the Uno Platform internal structure uses two layers of "bait-and-switch" of reference assemblies.
> 
> ### Publish-time switching
> 
> When building in `SkiaRenderer` node, an application compiles against "reference" versions of `Uno.UI`, `Uno.UI.Dispatching`, `Uno.UI.Composition`, `Uno` and `Uno.Foundation`. When the application is being packaged, `Uno.UI` and `Uno.UI.Composition` are switched to the Skia compatible versions. The `Uno.UI.Dispatching`, `Uno`, and `Uno.Foundation` are switched to their corresponding target platform versions.
> 
> By doing so, any use of the APIs provided by `Uno.UI.Dispatching`, `Uno`, and `Uno.Foundation` are automatically redirected to the proper platform support, for instance redirecting `GeoLocator` to use the proper APIs provided by the underlying platform.
> 
> ### Implications for iOS/Android class libraries
> 
> At this time, NuGet packages that are providing UI features on iOS/Android may not be consumed by both native and skia renderer projects under certain conditions.
> 
> The reason for this limitation is caused by the fact that Native renderers expose a different API set for `Microsoft.UI.Xaml` types, therefore causing incompatibility issues when apps use such a library.
> 
> Here are the different scenarios:
> 
> - Given a library that uses `net9.0-ios` or `net9.0-android`, which does not have the `SkiaRender` set as an `UnoFeature`, but uses platform conditional code with `#if` blocks:
> 
>   - This package is only be useable by both native and skia apps if it also provides a `net9.0` TFM. In this case, for a native app, nothing changes.
>   - For a `SkiaRenderer` enabled app, the `net9.0` variant of the library will be used and will not offer iOS/Android specific conditional code, but any code that uses Uno Platform provided APIs will work properly.
> 
> - Given a library that uses `net9.0-ios` or `net9.0-android`, which does have the `SkiaRender` set as an `UnoFeature`, but uses platform conditional code with `#if` blocks:
> 
>   - This package is only be useable by skia enabled apps.

Reviewer: ```suggestion
## Architecture

In order to accommodate the inclusion of Skia rendering for all platforms, the Uno Platform internal structure uses two layers of "bait-and-switch" of reference assemblies.

### Publish-time switching

When building in the `SkiaRenderer` node, an application compiles against "reference" versions of `Uno.UI`, `Uno.UI.Dispatching`, `Uno.UI.Composition`, `Uno`, and `Uno.Foundation`. When the application is being packaged, `Uno.UI` and `Uno.UI.Composition` are switched to the Skia-compatible versions. The `Uno.UI.Dispatching`, `Uno`, and `Uno.Foundation` are switched to their corresponding target platform versions.

By doing so, any use of the APIs provided by `Uno.UI.Dispatching`, `Uno`, and `Uno.Foundation` is automatically redirected to the proper platform support, for instance, redirecting `GeoLocator` to use the proper APIs provided by the underlying platform.

### Implications for iOS/Android class libraries

At this time, NuGet packages that provide UI features on iOS/Android may not be consumed by both native and Skia renderer projects under certain conditions.

The reason for this limitation is caused by the fact that Native renderers expose a different API set for `Microsoft.UI.Xaml` types, therefore causing incompatibility issues when apps use such a library.

Here are the different scenarios:

- Given a library that uses `net9.0-ios` or `net9.0-android`, which does not have the `SkiaRender` set as an `UnoFeature`, but uses platform conditional code with `#if` blocks:

  - This package is only usable by both native and Skia apps if it also provides a `net9.0` TFM. In this case, for a native app, nothing changes.
  - For a `SkiaRenderer` enabled app, the `net9.0` variant of the library will be used and will not offer iOS/Android specific conditional code, but any code that uses Uno Platform provided APIs will work properly.

- Given a library that uses `net9.0-ios` or `net9.0-android`, which does have the `SkiaRender` set as an `UnoFeature`, but uses platform conditional code with `#if` blocks:

  - This package is only usable by Skia-enabled apps.
```

---
uid: Uno.Skia.Rendering
---

# Using the Skia Rendering

> The Skia rendering engine is available for all platforms supported by Uno Platform. **Since Uno.Sdk 6.0+, it is the default rendering engine across all Uno Platform targets.** On **Windows (WinUI)**, **WebAssembly (Wasm)**, **Android**, and **iOS**, you can also choose to use the native rendering engine instead.

Reviewer 2: ```suggestion
The Skia rendering engine is available for all Uno Platform targets. **Since Uno.Sdk 6.0+, it is the default rendering engine.** On **WebAssembly (WASM)**, **Android**, and **iOS**, you can opt-in to use the native rendering engine instead. The **WinAppSDK** target is not provided by Uno Platform directly, so it only offers **native rendering mode**. 
```

Skia provides a consistent rendering pipeline across platforms and is often used to achieve improved performance or visual fidelity, particularly in scenarios where the default platform rendering may be lacking. Under the hood, the Uno Platform's Skia rendering engine is powered by SkiaSharp, a .NET binding to the Skia graphics library.

> [!Warning]

> > The Skia backend bypasses native controls and renders everything via SkiaCanvas, which can affect behavior for accessibility and platform-specific UI expectations.

Reviewer 2: ```suggestion
> The Skia backend bypasses native controls and renders everything onto a Skia canvas, which can affect behavior for accessibility and platform-specific UI expectations.
```

## Using Skia rendering for new apps

> You can use our [Visual Studio Wizard](xref:Uno.GettingStarted.UsingWizard) and enable the Skia rendering engine when creating a new project. This will set up the necessary MSBuild properties and references for you. You can find more details on how to use the wizard here: [Creating a new project](xref:Uno.GettingStarted.UsingWizard).

Author: Good point, this comes from the original commit where there weren't concrete details on the final state. I will adjust with the steps.

1. Next, enable Skia rendering by adding it to the `<UnoFeatures>` section in your `.csproj` file, like this:

```diff
<UnoFeatures>
  <!-- Existing features -->
+  SkiaRenderer;
</UnoFeatures>
```

#### Android

No additional setup is required for Skia on Android, but make sure to remove any legacy image loading or custom rendering logic that conflicts with Skia. For example, if you were previously using `UniversalImageLoader`, you can safely remove it:

```diff
 public Application(IntPtr javaReference, JniHandleOwnership transfer)
     : base(() => new App(), javaReference, transfer)
 {
-    ConfigureUniversalImageLoader();
 }

-private static void ConfigureUniversalImageLoader()
-{
-    var config = new ImageLoaderConfiguration.Builder(Context).Build();
-    ImageLoader.Instance.Init(config);
-    ImageSource.DefaultImageLoader = ImageLoader.Instance.LoadImageAsync;
-}
```

#### iOS

Replace the default `UIApplication.Main` startup with Skia’s iOS entry point:

```diff
+using Uno.UI.Runtime.Skia.AppleUIKit;

 public static class Program
 {
     public static void Main(string[] args)
     {
         App.InitializeLogging();
 
-        UIApplication.Main(args, null, typeof(App));

> +        var host = new AppleUIKitHost(() => new App());

Reviewer 2: This is now outdated, you should use `UnoPlatformBuilder` with `UseAppleUIKit()`

+        host.Run();
     }
 }
```

> #### WebAssembly (Wasm)

Reviewer 2: ```suggestion
#### WebAssembly (WASM)
```

Author: The correct one is actually Wasm - https://webassembly.org/ 😁

For WebAssembly, update the `Program` class to use the `WebAssemblyBrowserHost` provided by the Skia integration:

**Before:**

```csharp
public class Program
{
    private static App? _app;

    public static int Main(string[] args)
    {
        App.InitializeLogging();

        Microsoft.UI.Xaml.Application.Start(_ => _app = new App());

        return 0;
    }
}
```

**After:**

```csharp
using Uno.UI.Runtime.Skia.WebAssembly.Browser;

public class Program
{
    private static App? _app;

    public static async Task<int> Main(string[] args)
    {
        App.InitializeLogging();

>         var host = new WebAssemblyBrowserHost(() => _app = new App());

Reviewer 2: This is now outdated, you should use `UnoPlatformBuilder` with `UseWebAssembly()`

        await host.Run();

        return 0;
    }
}
```

## Benefits

> [!TIP]
> If you're building a custom drawing application, charts, or games in Uno Platform, Skia can offer more flexibility and uniform visuals across platforms.

- **Consistent visuals**: Skia ensures pixel-perfect rendering across all supported platforms, making it ideal for applications where precise control over appearance is critical.
- **Custom drawing**: Ideal for apps requiring advanced graphics, custom controls, or canvas-based rendering—Skia gives you low-level drawing access.
- **Unified rendering pipeline**: Unlike native rendering, which varies by platform, Skia uses a single rendering backend, reducing platform-specific inconsistencies.
- **Improved rendering performance on desktop**: On platforms like Linux/macOS, Skia is often faster and more efficient than native alternatives.
- **Better control over visual updates**: You can fine-tune repainting behavior for animations, games, or dynamic content using Skia’s immediate mode rendering.

> - **Smaller dependency surface**: By avoiding native UI components, Skia can simplify deployment—especially in environments like Wasm or containerized desktop apps.

Reviewer 2: ```suggestion
- **Smaller dependency surface**: By avoiding native UI components, Skia can simplify deployment—especially in environments like WASM or containerized desktop apps.
```

## Limitations

Using Skia rendering might have some limitations compared to native rendering. Some of the known limitations include:

> - **No native control integration**: Skia draws all UI elements as vector graphics, which means platform-native controls (like a native text box or button) are not used. This can affect accessibility and input handling.

Reviewer 2: There is an option to embed native controls

> - **Accessibility support**: Since Skia doesn't rely on native controls, accessibility tools (e.g., screen readers) may not function as expected.

Reviewer 2: Mention this is mainly a temporary limitation which we are working on to remove
